File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/PracticeCatalogSearchSupport.kt
package com.pillyliu.pinprofandroid.practice

import android.content.SharedPreferences
import com.pillyliu.pinprofandroid.library.PinballGame
import com.pillyliu.pinprofandroid.library.matchesSearchTokens
import com.pillyliu.pinprofandroid.library.normalizedSearchTokens
import java.util.Locale

internal const val KEY_PRACTICE_SEARCH_RECENTS = "practice-search-recents-v1"
internal const val PRACTICE_SEARCH_MAX_RECENTS = 20

internal enum class PracticeSearchTypeFilter(val rawValue: String, val label: String) {
    EM("em", "EM"),
    SS("ss", "SS"),
    LCD("lcd", "LCD"),
}

internal data class PracticeSearchResult(
    val canonicalGameId: String,
    val displayName: String,
    val manufacturer: String?,
    val year: Int?,
    val searchTokens: List<String>,
    val manufacturerTokens: List<String>,
    val categoryFields: List<String>,
    val yearFields: List<Int>,
)

internal fun practiceManufacturerSuggestions(
    options: List<String>,
    query: String,
): List<String> {
    val trimmed = query.trim()
    if (trimmed.isBlank()) return emptyList()
    val queryTokens = normalizedSearchTokens(trimmed)
    return options.filter { option ->
        matchesSearchTokens(queryTokens, normalizedSearchTokens(option))
    }.take(8)
}

internal fun filterPracticeSearchResults(
    results: List<PracticeSearchResult>,
    nameQuery: String,
    manufacturerQuery: String,
    yearQuery: String,
    selectedType: PracticeSearchTypeFilter?,
): List<PracticeSearchResult> {
    val targetYear = yearQuery.trim().toIntOrNull()
    val nameTokens = normalizedSearchTokens(nameQuery)
    val manufacturerTokens = normalizedSearchTokens(manufacturerQuery.trim())
    return results.filter { result ->
        val matchesName = nameTokens.isEmpty() || matchesSearchTokens(nameTokens, result.searchTokens)
        val matchesManufacturer = manufacturerTokens.isEmpty() || matchesSearchTokens(manufacturerTokens, result.manufacturerTokens)
        val matchesYear = targetYear == null || result.yearFields.contains(targetYear)
        val matchesType = selectedType == null || result.categoryFields.any { it == selectedType.rawValue }
        matchesName && matchesManufacturer && matchesYear && matchesType
    }
}

internal fun buildPracticeSearchResults(games: List<PinballGame>): List<PracticeSearchResult> {
    return games
        .groupBy { it.practiceKey }
        .mapNotNull { (canonicalGameId, groupedGames) ->
            if (canonicalGameId.isBlank()) return@mapNotNull null
            val displayName = practiceDisplayTitleForGames(groupedGames)
            val manufacturer = practiceSearchManufacturer(groupedGames)
            val year = practiceSearchYear(groupedGames)
            PracticeSearchResult(
                canonicalGameId = canonicalGameId,
                displayName = displayName,
                manufacturer = manufacturer,
                year = year,
                searchTokens = groupedGames.flatMap { game ->
                    listOf(
                        displayName,
                        game.name,
                        game.opdbName,
                        game.opdbShortname,
                        game.opdbCommonName,
                        game.opdbGroupShortname,
                    )
                }.flatMap { normalizedSearchTokens(it.orEmpty()) },
                manufacturerTokens = groupedGames
                    .mapNotNull { it.manufacturer?.trim()?.takeIf(String::isNotBlank) }
                    .flatMap(::normalizedSearchTokens),
                categoryFields = groupedGames.mapNotNull(::practiceSearchCategory).distinct(),
                yearFields = groupedGames.mapNotNull { game ->
                    game.year ?: practiceSearchYear(game.opdbManufactureDate)
                },
            )
        }
        .sortedWith(compareBy<PracticeSearchResult> { it.displayName.lowercase(Locale.US) }.thenBy { it.canonicalGameId.lowercase(Locale.US) })
}

internal fun buildPracticeSearchMetaLine(result: PracticeSearchResult): String {
    val parts = mutableListOf(result.manufacturer ?: "-")
    result.year?.let { parts += it.toString() }
    return parts.joinToString(" • ")
}

internal fun loadPracticeSearchRecents(prefs: SharedPreferences): List<String> {
    return prefs.getString(KEY_PRACTICE_SEARCH_RECENTS, null)
        ?.split('\n')
        ?.map { it.trim() }
        ?.filter { it.isNotBlank() }
        ?: emptyList()
}

internal fun rememberPracticeSearchRecent(
    prefs: SharedPreferences,
    canonicalGameId: String,
) {
    val trimmed = canonicalGameId.trim()
    if (trimmed.isBlank()) return
    val updated = buildList {
        add(trimmed)
        loadPracticeSearchRecents(prefs).forEach { existing ->
            if (!existing.equals(trimmed, ignoreCase = true)) {
                add(existing)
            }
        }
    }.take(PRACTICE_SEARCH_MAX_RECENTS)
    prefs.edit().putString(KEY_PRACTICE_SEARCH_RECENTS, updated.joinToString("\n")).apply()
}

private fun practiceSearchManufacturer(games: List<PinballGame>): String? {
    return games.mapNotNull { it.manufacturer?.trim()?.takeIf(String::isNotBlank) }
        .distinctBy { it.lowercase(Locale.US) }
        .sortedBy { it.lowercase(Locale.US) }
        .firstOrNull()
}

private fun practiceSearchYear(games: List<PinballGame>): Int? {
    return games.mapNotNull { it.year ?: practiceSearchYear(it.opdbManufactureDate) }
        .sorted()
        .firstOrNull()
}

private fun practiceSearchYear(raw: String?): Int? {
    val trimmed = raw?.trim().orEmpty()
    if (trimmed.length < 4) return null
    return trimmed.take(4).toIntOrNull()
}

private fun practiceSearchCategory(game: PinballGame): String? {
    return when {
        game.opdbDisplay == PracticeSearchTypeFilter.LCD.rawValue -> PracticeSearchTypeFilter.LCD.rawValue
        game.opdbType == PracticeSearchTypeFilter.EM.rawValue -> PracticeSearchTypeFilter.EM.rawValue
        game.opdbType == PracticeSearchTypeFilter.SS.rawValue -> PracticeSearchTypeFilter.SS.rawValue
        else -> null
    }
}

File: Pinball App 2/Pinball App 2/practice/PracticeCatalogSearchSupport.swift
import Foundation

enum PracticeGameTypeFilter: String, CaseIterable, Identifiable {
    case em
    case ss
    case lcd

    var id: String { rawValue }

    var label: String {
        switch self {
        case .em:
            return "EM"
        case .ss:
            return "SS"
        case .lcd:
            return "LCD"
        }
    }
}

struct PracticeGameSearchResult: Identifiable {
    let canonicalGameID: String
    let displayName: String
    let manufacturer: String?
    let year: Int?
    let searchTokens: [String]
    let manufacturerTokens: [String]
    let categoryFields: [String]
    let yearFields: [Int]

    var id: String { canonicalGameID }
}

struct PracticeGameSearchFilters {
    var nameQuery: String = ""
    var manufacturerQuery: String = ""
    var yearQuery: String = ""
    var selectedType: PracticeGameTypeFilter?

    var hasFilters: Bool {
        !nameQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !manufacturerQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !yearQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            selectedType != nil
    }
}

struct PracticeGameSearchIndex {
    static let empty = PracticeGameSearchIndex(
        results: [],
        resultByID: [:],
        manufacturerOptions: []
    )

    let results: [PracticeGameSearchResult]
    let resultByID: [String: PracticeGameSearchResult]
    let manufacturerOptions: [String]

    init(games: [PinballGame]) {
        let results = buildPracticeSearchResults(games)
        self.init(
            results: results,
            resultByID: Dictionary(uniqueKeysWithValues: results.map { ($0.canonicalGameID, $0) }),
            manufacturerOptions: Array(Set(results.compactMap(\.manufacturer)))
                .sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }
        )
    }

    func manufacturerSuggestions(for query: String) -> [String] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }
        let queryTokens = normalizedSearchTokens(trimmed)
        return manufacturerOptions.filter { option in
            matchesSearchTokens(queryTokens, haystackTokens: normalizedSearchTokens(option))
        }
        .prefix(8)
        .map { $0 }
    }

    func filteredResults(using filters: PracticeGameSearchFilters) -> [PracticeGameSearchResult] {
        let targetYear = Int(filters.yearQuery.trimmingCharacters(in: .whitespacesAndNewlines))
        let nameTokens = normalizedSearchTokens(filters.nameQuery)
        let manufacturerTokens = normalizedSearchTokens(
            filters.manufacturerQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        )

        return results.filter { result in
            let matchesName = nameTokens.isEmpty ||
                matchesSearchTokens(nameTokens, haystackTokens: result.searchTokens)
            let matchesManufacturer = manufacturerTokens.isEmpty ||
                matchesSearchTokens(manufacturerTokens, haystackTokens: result.manufacturerTokens)
            let matchesYear = targetYear == nil || result.yearFields.contains(targetYear ?? 0)
            let matchesType = filters.selectedType == nil ||
                result.categoryFields.contains(filters.selectedType?.rawValue ?? "")
            return matchesName && matchesManufacturer && matchesYear && matchesType
        }
    }

    func recentResults(for recentGameIDs: [String]) -> [PracticeGameSearchResult] {
        recentGameIDs.compactMap { resultByID[$0] }
    }

    func metaLine(for result: PracticeGameSearchResult) -> String {
        var parts: [String] = [result.manufacturer ?? "-"]
        if let year = result.year {
            parts.append(String(year))
        }
        return parts.joined(separator: " • ")
    }

    private init(
        results: [PracticeGameSearchResult],
        resultByID: [String: PracticeGameSearchResult],
        manufacturerOptions: [String]
    ) {
        self.results = results
        self.resultByID = resultByID
        self.manufacturerOptions = manufacturerOptions
    }
}

enum PracticeGameSearchRecentStore {
    private static let defaultsKey = "practice-game-search-recents-v1"
    private static let maxCount = 20

    static func load() -> [String] {
        let values = UserDefaults.standard.array(forKey: defaultsKey) as? [String] ?? []
        return values.filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
    }

    @discardableResult
    static func remember(_ canonicalGameID: String) -> [String] {
        let trimmed = canonicalGameID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return load() }
        var values = load().filter { $0 != trimmed }
        values.insert(trimmed, at: 0)
        let updated = Array(values.prefix(maxCount))
        UserDefaults.standard.set(updated, forKey: defaultsKey)
        return updated
    }
}

func buildPracticeSearchResults(_ games: [PinballGame]) -> [PracticeGameSearchResult] {
    Dictionary(grouping: games, by: \.canonicalPracticeKey)
        .compactMap { canonicalGameID, groupedGames in
            guard !canonicalGameID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
            let displayName = practiceDisplayTitleForGames(groupedGames)
            let manufacturer = practiceSearchManufacturer(for: groupedGames)
            let year = practiceSearchYear(for: groupedGames)
            let searchTokens = groupedGames
                .flatMap { game in
                    [
                        displayName,
                        game.name,
                        game.opdbName,
                        game.opdbShortname,
                        game.opdbCommonName,
                        game.opdbGroupShortname
                    ]
                }
                .flatMap { normalizedSearchTokens($0 ?? "") }
            let manufacturerTokens = groupedGames
                .compactMap(\.manufacturer)
                .flatMap(normalizedSearchTokens)
            let categoryFields = Array(Set(groupedGames.compactMap(practiceSearchCategory(for:))))
            let yearFields = groupedGames.compactMap { game in
                game.year ?? practiceSearchYear(from: game.opdbManufactureDate)
            }
            return PracticeGameSearchResult(
                canonicalGameID: canonicalGameID,
                displayName: displayName,
                manufacturer: manufacturer,
                year: year,
                searchTokens: searchTokens,
                manufacturerTokens: manufacturerTokens,
                categoryFields: categoryFields,
                yearFields: yearFields
            )
        }
        .sorted { lhs, rhs in
            let nameCompare = lhs.displayName.localizedCaseInsensitiveCompare(rhs.displayName)
            if nameCompare != .orderedSame {
                return nameCompare == .orderedAscending
            }
            return lhs.canonicalGameID.localizedCaseInsensitiveCompare(rhs.canonicalGameID) == .orderedAscending
        }
}

private func practiceSearchManufacturer(for games: [PinballGame]) -> String? {
    let normalized = games.compactMap { $0.manufacturer?.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
    return normalized.sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }.first
}

private func practiceSearchYear(for games: [PinballGame]) -> Int? {
    if let year = games.compactMap(\.year).sorted().first {
        return year
    }
    return games.compactMap { practiceSearchYear(from: $0.opdbManufactureDate) }.sorted().first
}

private func practiceSearchYear(from manufactureDate: String?) -> Int? {
    guard let prefix = manufactureDate?.prefix(4), prefix.count == 4 else { return nil }
    return Int(prefix)
}

private func practiceSearchCategory(for game: PinballGame) -> String? {
    if game.opdbDisplay == PracticeGameTypeFilter.lcd.rawValue {
        return PracticeGameTypeFilter.lcd.rawValue
    }
    if game.opdbType == PracticeGameTypeFilter.em.rawValue {
        return PracticeGameTypeFilter.em.rawValue
    }
    if game.opdbType == PracticeGameTypeFilter.ss.rawValue {
        return PracticeGameTypeFilter.ss.rawValue
    }
    return nil
}
